i#3544 riscv64: Enable dynamorio unit test compile

shiptux opened the pull request:
Make unit tests compile by disabling them all.
Next, enable one test at a time when it passes.

Add RISCV-V 64 support to suite/tests/linux/infinite.c, and clients/drcachesim/tests/raw2trace_unit_tests.cpp
Notice: There are still a large number of unimplemented in Dynamorio RISC-V 64, and it cannot be tested via QEMU

Issue #3544

shiptux commented:
@derekbruening

derekbruening commented on api/samples/CMakeLists.txt:
```diff
@@ -224,21 +224,25 @@ configure_DynamoRIO_global(OFF ON)
-add_sample_client(bbbuf       "bbbuf.c"         "drmgr;drreg;drx")
-add_sample_client(bbcount     "bbcount.c"       "drmgr;drreg;drx")
-add_sample_client(bbsize      "bbsize.c"        "drmgr;drx")
+if (NOT RISCV64) # FIXME i#3544: Port tests to RISC-V64
```
If the idea is to exclude absolutely everything in this file, could we have a single return() instead to avoid cluttering the file with if() statements?

derekbruening replied:
nit: s/FIXME/TODO/ 

Ditto below.

(Really old code uses FIXME for everything but new code should only use it for very broken things: prefer TODO for missing not yet implemented pieces and XXX for optional changes.)

derekbruening commented on api/samples/CMakeLists.txt:
```diff
@@ -224,21 +224,25 @@ configure_DynamoRIO_global(OFF ON)
-add_sample_client(opcode_count    "opcode_count.cpp"    "drmgr;drreg;drx;droption")
+if (NOT RISCV64) # FIXME i#3544: Port tests to RISC-V64
+  add_sample_client(div         "div.c"           "drmgr")
+  add_sample_client(empty       "empty.c"         "")
```
This empty client should definitely compile.

derekbruening commented on api/samples/CMakeLists.txt:
```diff
@@ -224,21 +224,25 @@ configure_DynamoRIO_global(OFF ON)
+if (NOT RISCV64) # FIXME i#3544: Port tests to RISC-V64
+  add_sample_client(bbbuf       "bbbuf.c"         "drmgr;drreg;drx")
+  add_sample_client(bbcount     "bbcount.c"       "drmgr;drreg;drx")
+  add_sample_client(bbsize      "bbsize.c"        "drmgr;drx")
```
These 3 should already compile: I don't see anything inside them that wouldn't compile for RISCV? What build error do you see with these?

derekbruening commented on core/ir/aarch64/codec.c:
```diff
@@ -6809,9 +6809,9 @@ decode_svemem_vec_imm5(uint enc, aarch64_reg_offset element_size, bool is_prefet
-        imm5, 0, mem_transfer, 0);
+    *opnd = opnd_create_vector_base_disp_aarch64(zn, DR_REG_NULL,
+                                                 get_opnd_size_from_offset(element_size),
+                                                 0, false, imm5, 0, mem_transfer, 0);
```
This change looks suspicious: please explain.

derekbruening replied:
I don't understand: you reverted an AArch64 commit in this PR?  Then that is a change from your PR (and likely a change we do not want to merge).

In the old days Github would not hide changes from merges: so a reviewer looking at what changed since the last review would see master branch changes if the author did an update mid-review.  If that is what is happening (because it is a personal repo?) then please do not update this PR branch until the review is finished.

derekbruening replied:
Looks like github is successfully hiding merge-from-master changes -- so this was all about an explicitly reverted commit; see below about clearly marking that and indicating what the plan is: only temporary to test the CI or actually proposing to permanently revert (in which case it should be a separate PR and should bring in the authors of the being-reverted PR).

derekbruening commented on core/ir/aarch64/codec.c:
```diff
@@ -6824,13 +6824,6 @@ encode_svemem_vec_imm5(uint enc, aarch64_reg_offset element_size, bool is_prefet
-            DR_EXTEND_UXTX ||
-        index_scaled || index_scale_amount != 0)
-        return false;
-
```
Again, removing code from the AArch64 decoder looks like a mistake?

shiptux replied:
Ditto below.

derekbruening commented on core/ir/aarch64/instr_create_api.h:
```diff
@@ -10966,13 +10966,13 @@
  *             For the  [\<Zn\>.S{, #\<imm\>}] variant:
  *             opnd_create_vector_base_disp_aarch64(Zn, DR_REG_NULL, OPSZ_4,
- *             DR_EXTEND_UXTX, 0, imm5, 0, opnd_size_from_bytes(dr_get_sve_vl() / 32), 0)
+ *             0, 0, imm5, 0, opnd_size_from_bytes(dr_get_sve_vl() / 32), 0)
```
Again, why is this RISCV PR deleting tokens from AArch64 decoder files?

shiptux replied:
Ditto below.

derekbruening commented on suite/tests/common/getretaddr.c:
```diff
@@ -112,6 +112,9 @@ GLOBAL_LABEL(FUNCNAME:)
         ldr      x30, [sp], #16
         ret
+# elif defined(RISCV64)
+        /*# FIXME i#3544: Port tests to RISC-V64*/
```
Please use standard syntax: no #; end punctuation and a space before the closing *.  Ditto below.

derekbruening commented on suite/tests/tools.c:
```diff
@@ -564,6 +564,9 @@ GLOBAL_LABEL(code_self_mod:)
         cbnz     w1, repeat1
         ret
+#elif defined(RISCV64)
+        /*FIXME i#3544: Port tests to RISC-V64*/
```
Please add spaces and end punctuation.  Ditto below.

derekbruening commented on suite/tests/tools.c:
```diff
@@ -597,32 +600,37 @@ GLOBAL_LABEL(FUNCNAME:)
         INC(REG_SCRATCH0)
         RETURN
         END_FUNC(FUNCNAME)
+#endif
```
It seems incorrect to exclude END_FUNC when DECLARE_FUNC is included.

shiptux commented:
I updated this PR to the latest master branch the day before yesterday. And then the CI won't be pasted. So I reverted the updated commit, this is the reason why some aarch64 decoder modifications in my PR. @derekbruening

shiptux commented:
I updated this PR to the latest master branch, and then I found that the CI could not pass. So I revert the updated commit, this is the reason why my PR has some modifications of the aarch64 decoder. @derekbruening

derekbruening commented on suite/tests/CMakeLists.txt:
```diff
@@ -1997,7 +2004,9 @@ if (X86)
   set(control_flags "nzcv")
 endif ()
+if (NOT RISCV64) # TODO i#3544: Port tests to RISC-V 64
 tobuild(common.${control_flags} common/${control_flags}.c)
```
Missing indentation

derekbruening commented on suite/tests/CMakeLists.txt:
```diff
@@ -2782,11 +2811,14 @@ if (X86)
-tochcon(client.drwrap-test.appdll textrel_shlib_t)
+if (NOT RISCV64) # TODO i#3544: Port tests to RISC-V 64
+  tobuild_appdll(client.drwrap-test client-interface/drwrap-test.c)
+  get_target_path_for_execution(drwrap_libpath client.drwrap-test.appdll "${location_suffix}")
```
Line too long: please wrap

derekbruening replied:
> I don't know if the too-long line you said that is:
> if (NOT RISCV64) # TODO i#3544: Port tests to RISC-V 64
> Or
> get_target_path_for_execution(drwrap_libpath client.drwrap-test.appdll "${location_suffix}")
> I have already moved # TODO i#3544: Port tests to RISC-V 64 to the second line.
> 
> -if (NOT RISCV64) # TODO i#3544: Port tests to RISC-V 64
> +if (NOT RISCV64)
> + # TODO i#3544: Port tests to RISC-V 64
>   tobuild_appdll(client.drwrap-test client-interface/drwrap-test.c)
>   get_target_path_for_execution(drwrap_libpath client.drwrap-test.appdll "${location_suffix}")

The comment is always immediately below the referenced line.  It is the get_target_path... which is over 80 chars.

derekbruening commented on suite/tests/CMakeLists.txt:
```diff
@@ -3574,17 +3613,19 @@ if (BUILD_CLIENTS)
     endif ()
 
+    if (NOT RISCV64)
     # Test offline traces.
```
Missing indentation on these 4 comment lines

derekbruening commented on suite/tests/CMakeLists.txt:
```diff
@@ -3810,7 +3851,7 @@ if (BUILD_CLIENTS)
     endif ()
-
+    if (NOT RISCV64)
     torunonly_drcacheoff(opcode_mix ${ci_shared_app} ""
```
Missing indentation on all these lines below, and missing TODO comment

derekbruening commented on suite/tests/CMakeLists.txt:
```diff
@@ -3574,17 +3613,19 @@ if (BUILD_CLIENTS)
+
+      torunonly_drcachesim(opcode_mix ${ci_shared_app}
+        "-instr_encodings -simulator_type opcode_mix" "")
+    endif(NOT RISCV64)
```
nit: missing space after endif

derekbruening commented on suite/tests/CMakeLists.txt:
```diff
@@ -4107,7 +4150,6 @@ if (BUILD_CLIENTS)
     if (AARCH64)
       torunonly_drcachesim(drstatecmp-fuzz drstatecmp-fuzz-app
         "-trace_after_instrs 10M -enable_drstatecmp" "" "")
-      set(tool.drcachesim.drstatecmp-fuzz_timeout 180)
```
This looks like a mistake: why was this line deleted?

derekbruening commented on suite/tests/CMakeLists.txt:
```diff
@@ -4650,16 +4698,21 @@ if (UNIX)
   endif ()
 
+  if (NOT RISCV64)
   # We pass -native_exec_list which causes us to call strcasecmp in libc,
```
Missing indentation on these lines below

derekbruening commented on suite/tests/CMakeLists.txt:
```diff
@@ -4680,22 +4734,28 @@ if (UNIX)
   endif ()
 
+  if (NOT RISCV64)
   # runs of other builds with custom DR options
```
Missing indentation on these 2 lines

derekbruening commented:
> I updated this PR to the latest master branch, and then I found that the CI could not pass. So I revert the updated commit, this is the reason why my PR has some modifications of the aarch64 decoder

Please add comments/labels explaining that in the future as otherwise it looks like an unintentional mistake when looking just at the diff.

derekbruening commented:
Looks good, except it looks like you've missed one comment about the too-long line: once that's fixed we can merge it in.

shiptux commented:
> Looks good, except it looks like you've missed one comment about the too-long line: once that's fixed we can merge it in.
 
I don't know if you said the too-long line is:
`if (NOT RISCV64) # TODO i#3544: Port tests to RISC-V 64`
Or
`get_target_path_for_execution(drwrap_libpath client.drwrap-test.appdll "${location_suffix}")`
I have already moved `# TODO i#3544: Port tests to RISC-V 64` to the second line.
```
if (NOT RISCV64) # TODO i#3544: Port tests to RISC-V 64
if (NOT RISCV64)
  # TODO i#3544: Port tests to RISC-V 64
  tobuild_appdll(client.drwrap-test client-interface/drwrap-test.c)
  get_target_path_for_execution(drwrap_libpath client.drwrap-test.appdll "${location_suffix}")
```

shiptux commented:
Sorry for replying too late. These changes are all done. @derekbruening
